### PR TITLE
WIP autoencoder

### LINKED ISSUE #8: add autoencoder

### PR DESCRIPTION
**Pull request by beckermr**
Thus PR adds an autoencoder to muffnn. The autoencoder supports a variety of metrics and is compatible with the scikit-learn transform interface. 

I additionally 

1. refactored the MLP classes into their own subpackage.
2. pinned conda to 4.0.5 to avoid conda bugs
3. upgraded TensorFlow to 0.11.0rc2

Closes #8

### CODE REVIEW

**Comment by beckermr** [comment]
@mheilman See the code in this PR for my first pass. I am still shaking the bugs out, but the MSE metric only works great.

**Comment by beckermr** [comment]
This is a conda bug. :(

**Comment by beckermr** [comment]
@mheilman Thus PR is ready. Unfortunately it ballooned to ~1500 after I wrote all of the tests. :(

I suggest we do it in chunks:

1. The refactoring of the MLP and base classes and the Dockerfile.
2. The autoencoder class itself.
3. The autoencoder tests.

**Comment by beckermr** [comment]
I am going to add a test for sparse inputs because why not. :)

**Comment by mheilman** [comment]
sorry for the slow response on this. i'm at a conference, so i probably won't be able to take a close look until next week.

**Comment by beckermr** [comment]
No worries! I am continuing to hack away on this in the mean time.

**Comment by beckermr** [comment]
@mheilman I rewrote the tests to eliminate a several hundred lines of duplicate code and removed a bunch of tests that duplicated each other for the most part.

File: muffnn/__init__.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
-from muffnn.base import MLPBaseEstimator
-from muffnn.mlp_classifier import MLPClassifier
-from muffnn.mlp_regressor import MLPRegressor
+from .mlp.mlp_classifier import MLPClassifier
+from .mlp.mlp_regressor import MLPRegressor
+from .autoencoder.autoencoder import Autoencoder
+from .core import TFPicklingBase
 
-__all__ = ['MLPBaseEstimator', 'MLPClassifier', 'MLPRegressor']
```
Comment (mheilman) [inline muffnn/__init__.py]: fwiw, i don't think you need `muffnn/autoencoder/__init__.py`

File: muffnn/autoencoder/autoencoder.py [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+
+import numpy as np
+import tensorflow as tf
+import scipy.sparse as sp
+
+from sklearn.base import TransformerMixin, BaseEstimator
+from sklearn.utils import check_array, check_random_state
+from sklearn.utils.validation import NotFittedError
+
+from ..core import TFPicklingBase, affine
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: I think it's best to avoid relative imports and just to `from muffnn.core ...` here.

File: muffnn/autoencoder/autoencoder.py [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+
+    Parameters
+    ----------
+    hidden_units : tuple or list, optional
+        A list of integers indicating the number of hidden layers and their
+        sizes. The list is reflected about the last element to form the
+        autoencoder. For example, if hidden_units is (256, 32, 16) then
+        the autoencoder has layers
+
+            [input_layer, 256, 32, 16, 32, 256, input_layer]
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: it seems like the second `input_layer` should be `output_layer`, or perhaps better yet, that both `input_layer`s should be `input_layer_size`.

File: muffnn/autoencoder/autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+    n_epochs : int, optional
+        The number of epochs (iterations through the training data) when
+        fitting.
+    dropout : float or None, optional
+        The dropout probability. If None, then dropout will not be used.
+        Note that dropout is not applied to the input layer.
+    activation : (callable, callable), optional
+        The activation functions for
+
+            (the bulk of the autoencoder and encoding layer, output layer).
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: unfinished here?

File: muffnn/autoencoder/autoencoder.py [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+
+        See `tensorflow.nn` for various options.
+    random_state: int, RandomState instance or None, optional
+        If int, the random number generator seed. If RandomState instance,
+        the random number generator itself. If None, then `np.random` will be
+        used.
+    learning_rate : float, optional
+        Learning rate for Adam.
+    metric : string, optional
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: i'd suggest either a) removing this kwarg or b) letting the user specify metrics for discrete variables in addition to ordinal variables.

File: muffnn/autoencoder/autoencoder.py [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+        output layer to be tensorflow.nn.sigmoid. Any dimensions indicated by
+        the options `categorical_begin`, `categorical_size` or
+        `discrete_indices` use a cross-entropy metric with
+        `tensorflow.nn.softmax` or `tensorflow.nn.sigmoid` output layer
+        activations.
+    discrete_indices : array-like, 1d, optional
+        Array of indices for which `tf.nn.sigmoid` will be used for the
+        output layer activation and the cross-entropy will be used as the
+        metric for the autoencoder.
+    categorical_begin : array-like, shape (n_categorial), optional
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: it seems like it'd be easier to just have one kwarg for categorical variables, of shape `(n_categorical, 2)` (there's a typo in docstring btw).

File: muffnn/autoencoder/autoencoder.py [hunk 7]
```diff
@@ -0,0 +1,682 @@
...
+        The batch size for learning and prediction. If there are fewer
+        examples than the batch size during fitting, then the the number of
+        examples will be used instead.
+    n_epochs : int, optional
+        The number of epochs (iterations through the training data) when
+        fitting.
+    dropout : float or None, optional
+        The dropout probability. If None, then dropout will not be used.
+        Note that dropout is not applied to the input layer.
+    activation : (callable, callable), optional
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: related to the comment below about `metric`, i'd suggest having this argument just be for the hidden layer.

With respect to the output layer, I'd suggest either a) just not letting the user specifying the output activation for now, or b) having a separate kwarg that would allow specification of activations for discrete as well as ordinal variables.

I'd go for the (a) and maybe add that functionality later if it's really needed.

sigmoid/softmax seems more obviously correct for discrete variables and less likely to need changing than MSE for ordinal/continuous data, of course, but it seems strange to have some the activation set in stone for some types of data but not others.

File: muffnn/autoencoder/autoencoder.py [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+            while start_idx < n_examples:
+                X_batch = \
+                    X[start_idx:min(start_idx + self.batch_size, n_examples)]
+                feed_dict = self._make_feed_dict(X_batch, training=False)
+                start_idx += self.batch_size
+                pred_batches.append(self._session.run(self._encoded_values,
+                                                      feed_dict=feed_dict))
+        return np.concatenate(pred_batches)
+
+    def inverse_transform(self, X, y=None, return_sparse=False):
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: What's the use case for this?

Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: If people do modeling on the encoded variables (say fitting a density estimate) that produces new ones, they can transform back to the input space.

Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: See the KDE digits example from scikit for a typical use case.

File: muffnn/autoencoder/autoencoder.py (hunk 2 above)
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Not following why...

File: muffnn/autoencoder/autoencoder.py (hunk 5 above)
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: What metrics besides cross-entropy w/ sigmoid/softmax do you expect to be useful for discrete variables?

File: muffnn/autoencoder/autoencoder.py (hunk 6 above)
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Sure I can make this change. I like the current structure since it is more explicit and you don't have to remember the order (i.e. is it `(begin, size)` or `(size, begin)`?)

File: muffnn/autoencoder/autoencoder.py (hunk 2 above)
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: it doesn't really matter much, but "explicit is better than implicit" and all that.  absolute imports are easier for the reader, imo

File: muffnn/autoencoder/autoencoder.py (hunk 4 above)
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: at a second glance, i see what this means.  I think it'd be cleared to say "A tuple of two callables for 1) the bulk of the autoencoder and encoding layer, and 2) the output layer."

File: muffnn/autoencoder/autoencoder.py (hunk 2 above)
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: hmmmm. I actually don't know this part of the python interpreter. Suppose I have two copies of muffnn, maybe one installed and another that is being run directly from a local dir. With an absolute import, which copy gets used?

File: muffnn/autoencoder/autoencoder.py (hunk 5 above)
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: i'm not sure if it really makes sense for autoencoders, but perhaps hinge loss (SVMs) or other loss functions that sklearn provides as [options for its SGD classifiers](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html).

File: muffnn/autoencoder/autoencoder.py (hunk 6 above)
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: `(begin, size)` seems like it'd make sense.  offset, length is pretty common (e.g., SQL).  one could also do `(begin, end)`.

It seems unusual to have 2 keyword arguments that are optional but require each other, particular when they have the same data type.

File: muffnn/autoencoder/autoencoder.py (hunk 2 above)
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: hmm... i *think* this issue is orthogonal to how packages get loaded from the filesystem.  iirc, if you're in a directory with a subdirectory that's a package, that'll take priority regardless of relative or absolute importing.

File: muffnn/autoencoder/autoencoder.py (hunk 5 above)
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: i think i understand your approach here, and i think it makes sense, but it seems like maybe there's a better way to expose the options.  how about this?

```
"""
An autoencoder using TensorFlow for discrete, continuous, or mixed data.

Parameters
----------
....
hidden_activation : callable, optional
    The activation function for the hidden layers.
    See `tensorflow.nn` for various options.
output_activation : callable, optional
    The activation function for the output layer.
    See `tensorflow.nn` for various options.
...
metric : string, optional
    Default metric for the autoencoder. Options are
        'mse' - mean square error
        'cross-entropy' - cross-entropy
    Note that this will be overridden for columns specified by `binary_indices`
    or `categorical_indices`
binary_indices : array-like, shape (n_binary,), optional
    Array of indices for which `tf.nn.sigmoid` will be used for the
    output layer activation and cross-entropy will be used in the loss
    function.
categorical_indices : array-like, shape (n_categorical, 2), optional
    An array where each row specifies a range of indices for a categorical
    variable that has been expanded to multiple indices (e.g., 1-hot encoded).
    The first column contains start indices. The second column contains
    lengths. For each range of indices, the output layer will use
    `tensorflow.nn.softmax` activation and the loss function will use
    cross-entropy.
...
"""
```

also, I don't think cross-entropy should force a sigmoid output layer.  One might want to use either softmax or sigmoid activation with cross-entropy, I think, depending on the problem.

File: muffnn/autoencoder/autoencoder.py (hunk 8 above)
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: ah ok.  it might be worth adding a sentence explaining the use case.  i think most users wouldn't need this and would be a bit confused by it

File: muffnn/__init__.py (hunk 1 above)
Comment (beckermr) [inline muffnn/__init__.py]: So after python 3.2, removing this file is OK I think. From the setuptools docs

"Anyway, find_packages() walks the target directory, filtering by inclusion patterns, and finds Python packages (any directory). On Python 3.2 and earlier, packages are only recognized if they include an __init__.py file. Finally, exclusion patterns are applied to remove matching packages."

I think we might want to retain it for all python 3 versions?

File: muffnn/autoencoder/autoencoder.py (hunk 5 above)
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: This looks good to me. I will adjust the code accordingly.

**Comment by mheilman** [comment]
lemme know whenever you're ready for me to take another look

**Comment by beckermr** [comment]
Yeah def. Will do this afternoon. Sorry for the delay.

**Comment by mheilman** [comment]
oh, no rush!  i just wanted to be clear i was done looking for the time being

**Comment by beckermr** [comment]
OK. This is fixed and ready for a second look @mheilman!

**Comment by mheilman** [comment]
ah, i just remembered your comment about doing this in chunks.  gonna do a first pass on the whole thing since i'm mostly done.

File: muffnn/autoencoder/tests/test_autoencoder.py [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+                                 hidden_units=(20, 20, 10, 10, 2),
+                                 dropout=0.05,
+                                 learning_rate=1e-3)
+
+    ae_score_nodropout = _check_ae(MSE_MAX_SCORE,
+                                   hidden_units=(20, 20, 10, 10, 2),
+                                   dropout=0.0,
+                                   learning_rate=1e-3)
+
+    assert ae_score_nodropout < ae_score_dropout, ("MSE with dropout should "
```
Comment (mheilman) [inline muffnn/autoencoder/tests/test_autoencoder.py]: theoretical question: it seems like this won't be the case in general, for training datasets and nonconvex models.  dropout could actually help avoid some local optimum and end up with a better representation of the training data, no?

File: Dockerfile [hunk 10]
```diff
@@ -24,14 +24,14 @@ RUN DEBIAN_FRONTEND=noninteractive apt-get update -y && \
 
 
 RUN echo 'export PATH=/opt/conda/bin:$PATH' > /etc/profile.d/conda.sh && \
-    wget --quiet https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh && \
-    /bin/bash /Miniconda3-latest-Linux-x86_64.sh -b -p /opt/conda && \
-    rm Miniconda3-latest-Linux-x86_64.sh && \
-    /opt/conda/bin/conda install --yes conda
+    wget --quiet https://repo.continuum.io/miniconda/Miniconda3-4.0.5-Linux-x86_64.sh && \
```
Comment (beckermr) [inline Dockerfile]: Cool. I think this version had some bug which had not been fixed yet. I will make the change and try it. I used 4.0.5 because I knew it worked.

File: muffnn/autoencoder/tests/test_autoencoder.py (hunk 9 above)
Comment (beckermr) [inline muffnn/autoencoder/tests/test_autoencoder.py]: I am not sure TBH.

Comment (beckermr) [inline muffnn/autoencoder/tests/test_autoencoder.py]: Empirically I have never seen this happen. You do get better on models on *test* data, but this is just eliminating overfitting.

File: muffnn/autoencoder/autoencoder.py [hunk 11]
```diff
@@ -0,0 +1,686 @@
...
+        ----------
+        X : numpy array or sparse matrix of shape [n_samples, n_features]
+            Encoded data to decode
+
+        return_sparse : bool, optional
+            If True, return a sparse matrix.
+
+        Returns
+        -------
+        Z : numpy array
```
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: To indicate that this is encoded data. I have seen this notation and using X would be confusing since that is an input.

File: muffnn/autoencoder/tests/test_autoencoder.py [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+def softmax_cross_entropy(ytrue, ypred):
+    """Compute the cross-entropy for a multi-class classifier.
+
+    This function computes
+
+        np.sum(-ytrue * np.log(ypred), axis=1)
+
+    It also handles 0's since they can produce NaNs.
+
+    See scikit.metrics.log_loss for a version with clipping of the
```
Comment (beckermr) [inline muffnn/autoencoder/tests/test_autoencoder.py]: Well my original thought was to match the TF computation to get the numerics to be the same. This is probably not happening, but I never went back and removed this.

File: muffnn/autoencoder/autoencoder.py [hunk 13]
```diff
@@ -0,0 +1,686 @@
...
+
+        self._categorical_msks = tf.placeholder(
+            tf.float32,
+            [None, None, self.input_layer_size_],
+            "categorical_msks")
+
+        # Fan in layers.
+        for i, layer_sz in enumerate(self.hidden_units):
+            # Don't dropout inputs.
+            if i > 0:
```
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: On second thought, I don't see the issue with only using dropout on the decoding part. The point here is to avoid overfitting and the idea of rescaling inputs (like one-hot encoded stuff), doesn't make sense to me. Using dropout only in the decoding part seems just fine to me.

File: muffnn/autoencoder/autoencoder.py [hunk 14]
```diff
@@ -0,0 +1,686 @@
...
+                        self._categorical_indices[:, 1]):
+                    scores += tf.nn.softmax_cross_entropy_with_logits(
+                        tf.slice(t, [begin, 0], [size, -1]),
+                        tf.slice(tf.transpose(self._input_values),
+                                 [begin, 0], [size, -1]),
+                        dim=0)
+
+                    # This one is painful. TensorFlow does not, AFAIK,
+                    # support assignments to Tensors that come out of
+                    # operations (it does support assignments to Variables).
```
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: The case condition in `tf.case` is a boolean scalar, so this won't get you around the mask thing.

File: muffnn/autoencoder/autoencoder.py [hunk 15]
```diff
@@ -0,0 +1,686 @@
...
+                scores += tf.reduce_sum(
+                    tf.nn.sigmoid_cross_entropy_with_logits(tsub, isub),
+                    reduction_indices=[0])
+
+                binary_msk = tf.transpose(self._binary_msk)
+                t = ((1.0 - binary_msk) * t +
+                     binary_msk * tf.nn.sigmoid(t))
+
+            # Anything left uses the default metric.
+            if self._default_indices is not None:
```
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Ahh it should be if the default ind set is empty. I will fix.

File: muffnn/autoencoder/autoencoder.py [hunk 16]
```diff
@@ -0,0 +1,686 @@
...
+                    if self.output_activation is not None:
+                        t = ((1.0 - default_msk) * t +
+                             default_msk * self.output_activation(t))
+                    tsub = tf.gather(t, self._default_indices)
+                    isub = tf.gather(tf.transpose(self._input_values),
+                                     self._default_indices)
+                    diff = isub - tsub
+                    scores += tf.reduce_sum(tf.square(diff),
+                                            reduction_indices=[0])
+                elif self.metric == 'cross-entropy':
```
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: You could, but then you are doing a bunch of extra work on your data for no reason (multiplying by masks, potentially in efficient gather operations over arrays, etc.). I think the multiple cases make sense here. Also, if you look carefully, the code itself is not duplicated. The code that does the mixed metric case is pretty different since it works on transposed stuff and has `tf.gather` calls.

File: muffnn/autoencoder/autoencoder.py [hunk 17]
```diff
@@ -0,0 +1,686 @@
...
+            # Undo the transpose here.
+            t = tf.transpose(t)
+
+        return t, scores
+
+    def _make_feed_dict(self, X, inverse=False, training=False):
+        # Make the dictionary mapping tensor placeholders to input data.
+
+        # Convert sparse inputs to dense.
+        if sp.issparse(X):
```
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Two things:

1. No, `check_array` will convert various input sparse formats to a single format, but does not do the sparse to dense conversion. 

2. My goal here was to only convert a sparse input to dense for each mini-batch instead of doing the whole data set at once in order to save memory. This is useful for giant bag-of-words representations, for example.

File: muffnn/autoencoder/autoencoder.py [hunk 18]
```diff
@@ -0,0 +1,686 @@
...
+        """
+
+        # Check that the input X is an array or sparse matrix.
+        # Each batch will be converted to dense on-the-fly.
+        # For sparse input, make the input a CSR matrix since it can be
+        # indexed by row.
+        X = check_array(X, accept_sparse=['csr'])
+
+        random_state = check_random_state(self.random_state)
+        assert self.batch_size > 0, "batch_size <= 0"
```
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Ack. This check_random call should only be done once, right?

File: muffnn/autoencoder/autoencoder.py [hunk 19]
```diff
@@ -0,0 +1,686 @@
...
+        inverses of each other for an autoencoder.
+
+        This method can be useful if further modeling on the encoded data
+        generates new encoded data vectors (e.g., generating new data by
+        fitting a density estimate to the encoded data).
+
+        Parameters
+        ----------
+        X : numpy array or sparse matrix of shape [n_samples, n_features]
+            Encoded data to decode
```
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Frack. This doc string is wrong.

File: muffnn/autoencoder/autoencoder.py (hunk 18 above)
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Yep. Moved this inside the is_fitted check.

File: muffnn/autoencoder/tests/test_autoencoder.py (hunk 12 above)
Comment (beckermr) [inline muffnn/autoencoder/tests/test_autoencoder.py]: I remember why now. The sklearn one sums over the samples auotmagically. I wanted to test the score of each sample.

File: muffnn/autoencoder/tests/test_autoencoder.py [hunk 20]
```diff
@@ -0,0 +1,539 @@
...
+                                   assert_almost_equal)
+
+from muffnn import Autoencoder
+
+
+_LOGGER = logging.getLogger(__name__)
+iris = load_iris()
+
+
+def sigmoid_cross_entropy(ytrue, ypred):
```
Comment (beckermr) [inline muffnn/autoencoder/tests/test_autoencoder.py]: Added. I also combined these two functions.

File: Dockerfile (hunk 10 above)
Comment (beckermr) [inline Dockerfile]: They seemed to have fixed this bug. 4.2.12 works fine.

**Comment by beckermr** [comment]
@mheilman This is ready for another look!

File: muffnn/autoencoder/autoencoder.py [hunk 21]
```diff
@@ -141,23 +141,22 @@ def _set_up_graph(self):
...
 
         self._categorical_msks = tf.placeholder(
-            tf.float32,
+            tf.bool,
             [None, None, self.input_layer_size_],
             "categorical_msks")
 
         # Fan in layers.
         for i, layer_sz in enumerate(self.hidden_units):
-            # Don't dropout inputs.
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: the docstring still says "Note that dropout is not applied to the input layer."

**Comment by mheilman** [comment]
hmmm... sorry i didn't think about this until now, but is there a check to make sure that indices aren't specified for both sigmoid and softmax, or for indices above the number of inputs?  input validation like that would be useful to help users avoid silly mistakes

**Comment by mheilman** [comment]
The docstring still says "Note that dropout is not applied to the input layer", which isn't the case anymore.

**Comment by beckermr** [comment]
Good call on index checking. I will move this into a separate routine.

**Comment by beckermr** [comment]
@mheilman I have addressed these comments. Back to you!

File: muffnn/autoencoder/autoencoder.py [hunk 22]
```diff
@@ -426,36 +425,100 @@ def __getstate__(self):
 
         # Add fitted attributes if the model has been fitted.
         if self._is_fitted:
-            state['input_layer_size_'] = self.input_layer_size_
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: hmm...  i made that comment, but now that i think about it, it's probably more efficient to set the values individually than make a separate dict, but this is fine.  in practice, it won't matter either way here

Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Yeah. We are really talking about microseconds here. :)

File: muffnn/autoencoder/autoencoder.py [hunk 23]
```diff
@@ -426,36 +425,100 @@ def __getstate__(self):
...
+        default_set = set()
+        if self._default_indices is not None:
+            default_set = set(self._default_indices.tolist())
+
+        if not sigmoid_set.isdisjoint(softmax_set):
+            raise ValueError("Sigmoid indices and softmax indices cannot"
+                             " overlap!")
+
+        if not sigmoid_set.isdisjoint(default_set):
+            raise ValueError("Sigmoid indices and default indices cannot"
```
Comment (mheilman) [inline muffnn/autoencoder/autoencoder.py]: i think this should be an assertion if it shouldn't depend on the user input.

actually, i think it'd be more common to leave out the assertion and instead have unit tests make sure this won't happen.  is testing that the default indices are computed already covered indirectly by the existing tests?

with my previous comment related to this, i was mainly concerned about the user passing in the same indices for softmax and sigmoid, which the check above and ValueError takes care of.

**Comment by mheilman** [comment]
only a couple minor comments.  thoughts?

File: muffnn/autoencoder/autoencoder.py (hunk 23 above)
Comment (beckermr) [inline muffnn/autoencoder/autoencoder.py]: Yeah, these extra checks are def covered by the unit tests. I will adjust the code.

**Comment by beckermr** [comment]
OK @mheilman. I think we are really close now!

**Comment by mheilman** [comment]
```
    __    ______________  ___
   / /   / ____/_  __/  |/  /
  / /   / / __  / / / /|_/ / 
 / /___/ /_/ / / / / /  / /  
/_____/\____/ /_/ /_/  /_/                                
```
(LGTM)

**Comment by mheilman** [comment]
my only thought, which would be for another PR, is that it might be nice to have options for penalties on the hidden layer (a la [this paper](http://web.stanford.edu/class/cs294a/sparseAutoencoder.pdf)).
